cpp/include/raft/core/device_resources_snmg.hpp

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/core/device_resources.hpp>
+
+#include <nccl.h>
+#include <omp.h>
+
+#include <memory>
+#include <vector>
+
+/**
+ * @brief Error checking macro for NCCL runtime API functions.
+ *
+ * Invokes a NCCL runtime API function call, if the call does not return ncclSuccess, throws an
+ * exception detailing the NCCL error that occurred
+ */
+#define RAFT_NCCL_TRY(call)                        \
+  do {                                             \
+    ncclResult_t const status = (call);            \
+    if (ncclSuccess != status) {                   \
+      std::string msg{};                           \
+      SET_ERROR_MSG(msg,                           \
+                    "NCCL error encountered at: ", \
+                    "call='%s', Reason=%d:%s",     \
+                    #call,                         \
+                    status,                        \
+                    ncclGetErrorString(status));   \
+      throw raft::logic_error(msg);                \
+    }                                              \
+  } while (0);
+
+namespace raft {
+
+/**
+ * @brief SNMG resource container object that stores a NCCL clique and
+ * all necessary resources used for calling device functions, cuda kernels,
+ * libraries and/or NCCL communications on each GPU
+ *
+ * Note that the NCCL communications sometimes require
```

Comment:
Are we able to expand upon this a little bit more here? Are there specific things the user should be aware of or avoid? Are the SNMG-copable algorithms responsible for this? Are there any threading resources needed in order to make this possible? Is it 1 thread per gpu? Do we need to provide a thread pool here too?

Reply:
Sure, I will expand a bit on this mentioning more specifically the blocking NCCL calls such as send&receive. NCCL clique destruction also require multiple threads, but this is managed in the destructor method. Yes, SNMG-capable algorithms will be responsible for this. OpenMP can be used for this, but other libs will work too. I think it is preferable to stay agnostic to the lib used. A user could therefore create his own thread pool or simply use OpenMP loops. This do not necessarily have to be 1 thread per GPU, and if exchanges are managed separately, they could even take place in a single thread thanks to the use of a NCCL group. But, since in most cases exchanges happen in the middle of other computations, and computations are for the most part similar for every GPU, it is convenient to have 1 thread per GPU. That's why I felt like the user should be made aware of his responsibility for managing NCCL comms accross threads.